### PR TITLE
Feature branch

### PR DESCRIPTION
**Pull request by NikalaiSamtsevich**

### CODE REVIEW

File: src/main/kotlin/com/libraryapp/keycloakauthservice/api/UserApi.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        @RequestParam(required = false) username: String?
+    ): List<UserRepresentation> {
+        return userService.getUsers(
+            email,
+            idpAlias,
+            idpUserId,
+            max,
+            username
+        )
+    }
```
Comment (psidarovich) [inline src/main/kotlin/com/libraryapp/keycloakauthservice/api/UserApi.kt]: Check expression functions in Kotlin. Here and futher

File: src/main/kotlin/com/libraryapp/keycloakauthservice/model/NewUserRecord.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    @field:Schema(description = "Email address of the new user", example = "johndoe@example.com")
+    val email: String?,
+
+    @field:NotBlank(message = "First name is required")
+    @field:Schema(description = "First name of the new user", example = "John")
+    val firstname: String?,
+
+    @field:NotBlank(message = "Last name is required")
+    @field:Schema(description = "Last name of the new user", example = "Doe")
+    val lastname: String?
```
Comment (psidarovich) [inline src/main/kotlin/com/libraryapp/keycloakauthservice/model/NewUserRecord.kt]: Are you sure, that your fields can be nullable?

File: src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/KeycloakAdminRestApi.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                keycloakConfig.clientSecret,
+                keycloakConfig.adminGrantType,
+            )
+            return webClient.post()
+                .uri(keycloakConfig.tokenUri)
+                .body(body)
+                .retrieve()
+                .awaitBody<AccessTokenResponse>()
+        } catch (e: Exception) {
+            println("Error occurred: ${e.message}")
```
Comment (psidarovich) [inline src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/KeycloakAdminRestApi.kt]: Let's use logger for printing ex messages

File: src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/RoleServiceImpl.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        webClient.post()
+            .uri("${keycloakConfig.userEndpoint}/$userId/role-mappings/realm")
+            .bodyValue(listOf(role))
+            .header("Authorization", "Bearer ${tokenResp.token}")
+            .retrieve()
+            .onStatus({ status -> status.isError }, { resp ->
+                resp.bodyToMono<String>()
+                    .flatMap { body -> Mono.error(ResponseStatusException(resp.statusCode(), body)) }
+            })
+            .awaitBodyOrNull<Unit>()
```
Comment (psidarovich) [inline src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/RoleServiceImpl.kt:35]: I personally don't remember, but are you sure, that calling this method will suspend function?

File: src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/UserServiceImpl.kt [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    fun userDtoToRepresentation(newUserRecord: NewUserRecord): UserRepresentation {
+        val userRepresentation: UserRepresentation = UserRepresentation()
+        userRepresentation.isEnabled = true;
+        userRepresentation.setUsername(newUserRecord.username);
+        userRepresentation.setEmail(newUserRecord.email);
+        userRepresentation.setFirstName(newUserRecord.username);
+        userRepresentation.setLastName(newUserRecord.lastname);
+        userRepresentation.setCreatedTimestamp(System.currentTimeMillis());
+        userRepresentation.isEmailVerified = false;
```
Comment (psidarovich) [inline src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/UserServiceImpl.kt]: Check scope function .apply() for better readability

**Review by psidarovich — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/RoleServiceImpl.kt (hunk 4 above)
Comment (NikalaiSamtsevich) [reply src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/RoleServiceImpl.kt:35]: Yes, if you're referring to awaitBodyOrNull.

File: src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/UserServiceImpl.kt [hunk 6]
```diff
@@ -77,18 +77,20 @@ class UserServiceImpl(
...
-        val credentialRepresentation: CredentialRepresentation = CredentialRepresentation()
-        credentialRepresentation.value = newUserRecord.password;
-        credentialRepresentation.type = CredentialRepresentation.PASSWORD;
+        val userRepresentation: UserRepresentation = UserRepresentation().apply {
+            isEnabled = true;
+            setUsername(newUserRecord.username);
+            setEmail(newUserRecord.email);
+            setFirstName(newUserRecord.username);
+            setLastName(newUserRecord.lastname);
+            setCreatedTimestamp(System.currentTimeMillis());
```
Comment (psidarovich) [inline src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/UserServiceImpl.kt]: Can we use Kotlin setters? I mean to use more direct approach:

`username = newUserRecord.username`

Comment (psidarovich) [reply src/main/kotlin/com/libraryapp/keycloakauthservice/service/impl/UserServiceImpl.kt]: Also remove `;`

**Review by psidarovich — CHANGES_REQUESTED** [review]

**Review by psidarovich — APPROVED** [review]

**Pull request merged**